include/SDL3/SDL_dialog.h

```diff
@@ -0,0 +1,190 @@
+  freely, subject to the following restrictions:
+
+  1. The origin of this software must not be misrepresented; you must not
+     claim that you wrote the original software. If you use this software
+     in a product, an acknowledgment in the product documentation would be
+     appreciated but is not required.
+  2. Altered source versions must be plainly marked as such, and must not be
+     misrepresented as being the original software.
+  3. This notice may not be removed or altered from any source distribution.
+*/
+
+#ifndef SDL_dialog_h_
+#define SDL_dialog_h_
+
+#include <SDL3/SDL_video.h>
+
+#include <SDL3/SDL_begin_code.h>
+/* Set up for C function definitions, even when using C++ */
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * An entry for filters for file dialogs.
+ *
+ * `name` is a user-readable label for the filter (for example, "Office document").
+ *
+ * `pattern` is a semicolon-separated list of file extensions (for example,
+ * "doc;docx").
+ *
+ * \sa SDL_DialogFileCallback
+ * \sa SDL_ShowOpenFileDialog
+ * \sa SDL_ShowSaveFileDialog
+ * \sa SDL_ShowOpenFolderDialog
+ */
+typedef struct
+{
+    const char* name;
+    const char* pattern;
+} SDL_DialogFileFilter;
+
+/**
+ * Callback used by file dialog functions.
+ *
+ * The specific usage is described in each function.
+ *
+ * If @p filelist is...
+ * - `NULL`, an error occured. Details can be obtained with SDL_GetError().
+ * - A pointer to `NULL`, the user either didn't choose any file or canceled
+ *   the dialog.
+ * - A pointer to non-`NULL`, the user chose one or more files. The argument is
+ *   a null-terminated list of pointers to C strings, each containing a path.
+ *
+ * The @p filelist argument does not need to be freed; it will automatically be
+ * freed when the callback returns.
+ *
+ * The @p filter argument is the index of the filter that was selected, or one
+ * more than the size of the list (therefore the index of the terminating NULL
+ * entry) if no filter was selected, or -1 if the platform or method doesn't
+ * support fetching the selected filter or if an error occured.
+ *
+ * \sa SDL_DialogFileFilter
+ * \sa SDL_ShowOpenFileDialog
+ * \sa SDL_ShowSaveFileDialog
+ * \sa SDL_ShowOpenFolderDialog
+ */
+typedef void(*SDL_DialogFileCallback)(void* userdata, const char* const* filelist, int filter);
+
+/**
+ * Displays a dialog that lets the user select a file on their filesystem.
+ *
+ * This function should only be invoked from the main thread.
+ *
+ * This is an asynchronous function; it will return immediately, and the result
+ * will be passed to the callback.
+ *
+ * The callback will be invoked with a null-terminated list of files the user
+ * chose. The list will be empty if the user canceled the dialog, and it will
+ * be NULL if an error occured.
+ *
+ * Note that the callback may be called from a different thread than the one
+ * the function was invoked on.
+ *
+ * If an error occurs, the error can be obtained with SDL_GetError().
```

Comment:
In general we require that a function return an error before SDL_GetError() can be called. You should probably remove this line and note instead that the file callback will be called with NULL.

Reply:
Now that I think about it, if the functions are asynchronous, `SDL_SetError()` might be called while some other code is busy doing other SDL code, potentially overwriting other error messages.

Is there another way to transmit error messages to the developers? Based on my personal experience, I feel like an absence of error messages can be a great source of frustration, especially given that some errors are not obvious (such as Zenity's arbitrary 2048 character limit for filters).

Reply:
You can use SDL_LogError(), but SDL_SetError() is thread local, so if you call that before calling the callback with NULL to signify the error, that should be fine.

Reply:
I removed the line entirely; it is already documented two paragraphs earlier that the list of files will be `NULL` if an error occurs.